### DISCUSSION TITLE
os.VfsLfs2.mkfs(sdcard) TypeError: function takes 3 positional arguments but 4 were given

### DISCUSSION THREAD

**Discussion by DavesCodeMusings** [micropython/micropython#13111] [Using MicroPython] [answered]
I am trying to format a microSD card with LittleFS and I'm getting an error.

Here's what I'm doing and the resulting error:

```
>>> import os
>>> os.VfsLfs2.mkfs(sdcard)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: function takes 3 positional arguments but 4 were given
```

The card is 32G, currently formatted as FAT, and plugged into an SPI connected breakout board. I can access it from /sd by putting this in my boot.py:

```
SD_MOUNT = "/sd"
print(f"Mounting micro SD card on: {SD_MOUNT}")
sdcard = SDCard(slot=3, sck=Pin(12), miso=Pin(13), mosi=Pin(11), cs=Pin(10))
vfs = VfsFat(sdcard)
mount(vfs, SD_MOUNT)
```

I can successfully `listdir("/sd")`, so I don't believe it's a problem with hardware.

After commenting out the lines with `VfsFat` and `mount`, I'm then running the commands at the REPL prompt as I've shown above.

Effectively, this:
```
>>> sdcard = SDCard(slot=3, sck=Pin(12), miso=Pin(13), mosi=Pin(11), cs=Pin(10))
>>> os.VfsLfs2.mkfs(sdcard)
```

I was assuming _sdcard_ is a block device, but I'm thinking maybe not.

I also tried this approach, trying to mirror what worked when mounting as FAT:

```
>>> sdcard = SDCard(slot=3, sck=Pin(12), miso=Pin(13), mosi=Pin(11), cs=Pin(10))
>>> vfs = os.VfsLfs2(sdcard)
>>> os.VfsLfs2.mkfs(sdcard)
```

Same result... _TypeError: function takes 3 positional arguments but 4 were given_

Can anyone tell me what I'm doing wrong? Is LittleFS just not meant for SD cards?

**Comment by jimmo** [comment, marked as answer]
> Is LittleFS just not meant for SD cards?

The sdcard driver uses the "simple interface" while littlefs only implements the "extended interface".

https://docs.micropython.org/en/latest/library/os.html#simple-and-extended-interface

We need to update sdcard.py to support the extended interface (and really...deprecate the simple interface)

**Reply by DavesCodeMusings** [reply]
I learned something new today. Thank you.

**Reply by jornamon** [reply]
Hi @DavesCodeMusings  and @jimmo !

Just in case you're still interested in the topic, I wanted to share that I've successfully integrated LittleFS with an SD card by adapting the existing driver to implment the block device extended interface. You can check out my work here: https://github.com/jornamon/mp-sdcard-littleFS.

Although LittleFS may not be the ideal choice for SD cards, the cache implemented in the driver makes it quite usable. Additionally, by tweaking the cache parameters, you might even see a performance boost while using FAT in certain scenarios. I'd love to hear your thoughts or any feedback you might have!

**Reply by DavesCodeMusings** [reply]
Oh, nice! I just skimmed your README for the project. It looks like this could work for what I was trying to do. I have an [asyncio FTP server](https://github.com/DavesCodeMusings/ftpdlite/) I whipped up, and the resistance to data corruption is what enticed me to try LittleFS in the first place. It would be interesting to see what kind of performance difference there is with a FAT-based SD card vs. a LittleFS-based SD card. When our lovely early spring weather turns back to clouds and cold (which it inevitably does) I'll sit down and give this another go.

**Reply by jornamon** [reply]
Oh! I saw you project some time ago. Nice! I didn't realize you were the same person. It's interesting to learn why you were trying to run LFS on a SD card. 

The driver will definitely allow you to use LFS. The performance will be worse than FAT, as you can see in the [test results](https://github.com/jornamon/mp-sdcard-littleFS/blob/main/TEST_RESULTS.md), but given that there are other limitations on microcontrollers, maybe access time to SD card is not the limiting factor here. Just beware that bigger cache sizes may improve overall performance but may also wipe out some of the resistance to data corruption. I don't know how much because I largely ignore the inner workings of LFS, but it's a possibility.

Let me know if you end up trying it.
